### PR TITLE
warn about registering customer service /5

### LINKED ISSUE #29152: Problem details
Add to [Problem details](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/error-handling?view=aspnetcore-8.0#problem-details)

See [this comment](https://github.com/dotnet/aspnetcore/issues/43261#issuecomment-1532480434)

> > Any writer registered before the call to the `AddProblemDetails `will be evaluated before the `DefaultProblemDetailsWrapper`
> 
> I had to register my custom `IProblemDetailsWriter` before the call to `AddRazorPages` AND the call to `AddProblemDetails` for it to be evaluated before the `DefaultProblemDetailsWriter`.

@zwoolli  If I am not mistaken `AddRazorPages` probably calls `AddMVCCore` (need to double check) that register a default writer for API Controllers. I believe your case the problem is not the `DefaultProblemDetailsWriter` but the `DefaultApiProblemDetailsWriter`. In both cases you are right and `AddProblemDetails` must be called before.

@Rick-Anderson maybe we could add something about it in the docs if not there yet. 

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: 0e03fdf4-b19b-17c3-04f4-336ab6d2ee06
* Version Independent ID: 38515dfb-91a5-b395-db9d-084bbaf095c8
* Content: [Handle errors in ASP.NET Core](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/error-handling?view=aspnetcore-8.0)
* Content Source: [aspnetcore/fundamentals/error-handling.md](https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/fundamentals/error-handling.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-fundamentals**
* GitHub Login: @tdykstra
* Microsoft Alias: **riande**

---
[Associated WorkItem - 88397](https://dev.azure.com/msft-skilling/Content/_workitems/edit/88397)

### PR DESCRIPTION
**Pull request by Rick-Anderson**
Fixes #29152

Bruno, can't we get an [ASPnnnn](https://learn.microsoft.com/en-us/aspnet/core/diagnostics/code-analysis?view=aspnetcore-8.0) code analysis to detect this?

[Review URL internal only](https://review.learn.microsoft.com/en-us/aspnet/core/fundamentals/error-handling?branch=pr-en-us-29221&view=aspnetcore-8.0#customize-problem-details)

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [aspnetcore/fundamentals/error-handling.md](https://github.com/dotnet/AspNetCore.Docs/blob/74a8ff4f110e6a21a8b6d194ed1a8ef26ffba339/aspnetcore/fundamentals/error-handling.md) | [Handle errors in ASP.NET Core](https://review.learn.microsoft.com/en-us/aspnet/core/fundamentals/error-handling?branch=pr-en-us-29221) |

### CODE REVIEW

**Comment by brunolins16** [comment]
> Bruno, can't we get an [ASPnnnn](https://learn.microsoft.com/en-us/aspnet/core/diagnostics/code-analysis?view=aspnetcore-8.0) code analysis to detect this?

@mitchdenny @captainsafia What do you think about this?

**Comment by Rick-Anderson** [comment]
cc @david-acker

File: aspnetcore/fundamentals/error-handling.md [hunk 1]
```diff
@@ -247,6 +247,8 @@ The automatic creation of a `ProblemDetails` can be customized using any of the
 2. Use a custom [`IProblemDetailsWriter`](#custom-iproblemdetailswriter)
 3. Call the [`IProblemDetailsService` in a middleware](#problem-details-from-middleware)
 
+***Note:*** When using a custom `IProblemDetailsWriter`, the custom `IProblemDetailsWriter` must be registered before calling <xref:Microsoft.Extensions.DependencyInjection.MvcServiceCollectionExtensions.AddRazorPages%2A>, <xref:Microsoft.Extensions.DependencyInjection.MvcServiceCollectionExtensions.AddControllers%2A>, or <xref:Microsoft.Extensions.DependencyInjection.MvcServiceCollectionExtensions.AddControllersWithViews%2A>.
```
Comment (brunolins16) [inline aspnetcore/fundamentals/error-handling.md:250]: What do you think about adding it under https://review.learn.microsoft.com/en-us/aspnet/core/fundamentals/error-handling?branch=pr-en-us-29221&view=aspnetcore-8.0#custom-iproblemdetailswriter, call out that it must be registered before AddProblemDetails as well and maybe add an example of registration?

**Comment by david-acker** [comment]
It seems like adding an analyzer to detect this would be straightforward. The one for ASP0001 serves a similar purpose.

**Comment by Rick-Anderson** [comment]
> It seems like adding an analyzer to detect this would be straightforward. The one for ASP0001 serves a similar purpose.

@david-acker  can you review https://github.com/dotnet/aspnetcore/issues/48178 and suggest any improvements?

**Comment by brunolins16** [comment]
> 

@Rick-Anderson the team usually create analyzer issues using this template https://github.com/dotnet/aspnetcore/issues/new?assignees=&labels=api-suggestion%2C+analyzer&projects=&template=60_analyzer_proposal.md&title=

**Comment by david-acker** [comment]
dotnet/aspnetcore#48180 LGTM!

**Comment by captainsafia** [comment]
@david-acker Thanks David! You rock! 🎸

**Pull request merged**